activerecord/test/cases/inheritance_test.rb

```diff
@@ -315,6 +315,13 @@ def test_new_without_storing_full_sti_class
     end
   end
 
+  def test_new_when_passing_class_to_type_when_store_attributes_are_present_on_sti_class
+    without_store_full_sti_class do
+      item = Company.new(type: Firm)
```

Comment:
I don't find specs for it, so no, not really sure we do. But looks like it has been supported without it for a long time, the typecasting takes care of it.
I agree, we will need to check if we should allow this, before actually adding this. cc @sgrif